### ISSUE TITLE
cpptools-srv hangs on ranges::views::enumerate inside find_template_class_full

### ISSUE THREAD

**Issue by intractabilis** [microsoft/vscode-cpptools#7455]
Bug type: Language Service

- OS and Version: Manjaro Ornara 21.0.3
- VS Code Version: 1.55.2
- C/C++ Extension Version: 1.3.1

Consider the following MWE:
```
#include <vector>

#include <range/v3/view/enumerate.hpp>

int main(int argc, char* argv[]) {
    std::vector<double> a;
    for (auto &&x : ranges::views::enumerate(a)) {
    }
}
```
With the following `CMakeLists.txt`
```
cmake_minimum_required(VERSION 3.20)
project(mwe)
find_package(range-v3 CONFIG REQUIRED)
add_executable(mwe mwe.cpp)
set_property(TARGET mwe PROPERTY CXX_STANDARD 20)
```

If you open it in VS Code and configure with CMake Tools, `cpptools-srv` process will hang indefinitely with one of the threads occupying 100% of CPU. Even after you exit from VS Code, both `cpptools` and `cpptools-srv` processes will remain.

The stack trace of the busy thread of `cpptools-srv` looks like this:
```
#0  equiv_template_arg_lists(a_template_arg*, a_template_arg*, int)() at :0
#1  hash_find(a_hash_table*, void*, int)() at :0
#2  find_template_class_full(a_symbol*, a_template_arg**, int, a_symbol*, int, int, int, int)() at :0
#3  find_template_class(a_symbol*, a_template_arg**, int, a_symbol*, int, int, int)() at :0
#4  coalesce_template_class_reference(a_symbol*, long long, int*)() at :0
#5  coalesce_template_id(a_symbol*, a_token_kind, unsigned int, long long, int, int*)() at :0
#6  f_is_generalized_identifier_start(long long, a_type*)() at :0
#7  decl_specifiers(unsigned long, a_decl_parse_state*, a_decl_pos_block*)() at :0
#8  type_name_full(a_decl_parse_state*)() at :0
#9  type_name(a_type**)() at :0
#10  scan_builtin_operation_arg(a_rescan_control_block*, an_il_entry_kind)() at :0
#11  scan_call_like_builtin_operation(a_rescan_control_block*, a_builtin_operation_kind_tag, a_type*, an_il_entry_kind, an_il_entry_kind, int, an_operand*)() at :0
#12  scan_binary_type_trait_helper(a_rescan_control_block*, an_operand*)() at :0
#13  scan_expr_full(an_operand*, an_operand*, int, int)() at :0
#14  scan_expr_full(an_operand*, an_operand*, int, int)() at :0
#15  scan_template_argument_constant_expression(a_type*, a_constant*, a_template_arg*, a_template_param*)() at :0
#16  scan_template_argument_list(a_symbol*, int, int*, long long, long*)() at :0
#17  coalesce_template_class_reference(a_symbol*, long long, int*)() at :0
#18  coalesce_template_id(a_symbol*, a_token_kind, unsigned int, long long, int, int*)() at :0
#19  f_is_generalized_identifier_start(long long, a_type*)() at :0
#20  scan_base_specifier_list(a_class_def_state*)() at :0
#21  scan_class_definition(a_type*, a_decl_parse_state*, int, int, int, int, int, int, a_template*, a_decl_pos_block*)() at :0
#22  f_instantiate_template_class_full(a_type*, unsigned int*, unsigned int*)() at :0
#23  scan_base_specifier_list(a_class_def_state*)() at :0
#24  scan_class_definition(a_type*, a_decl_parse_state*, int, int, int, int, int, int, a_template*, a_decl_pos_block*)() at :0
#25  f_instantiate_template_class_full(a_type*, unsigned int*, unsigned int*)() at :0
#26  scan_base_specifier_list(a_class_def_state*)() at :0
#27  scan_class_definition(a_type*, a_decl_parse_state*, int, int, int, int, int, int, a_template*, a_decl_pos_block*)() at :0
#28  f_instantiate_template_class_full(a_type*, unsigned int*, unsigned int*)() at :0
...
```
In the above the sequence of `scan_base_specifier_list`, `scan_class_definition`, and `f_instantiate_template_class_full` repeats down forever, I checked several hundreds stack frames, it still continued. Calls above `find_template_class_full`  change, so I assume that the hang is inside `find_template_class_full`.

Workspace `settings.json`:
```
{
    "cmake.configureOnOpen": true,
    "C_Cpp.default.configurationProvider": "ms-vscode.cmake-tools",
    "C_Cpp.loggingLevel": "Debug"
}
```

The log of the language server is attached.
[cpptools.log](https://github.com/microsoft/vscode-cpptools/files/6411067/cpptools.log)

**Comment by sean-mcmanus** [comment]
Thanks for reporting this. I've filed an issue on our shared VS parser at https://developercommunity.visualstudio.com/t/C20-IntelliSense-vcpkgsrvexe-uses-e/1412745

**Comment by intractabilis** [comment]
Thanks! Do you think the issues like this could have been avoided if you used Clang front-end instead of EDG?

**Comment by sean-mcmanus** [comment]
For users of the clang compiler -- however, our extension supports gcc, cl.exe, nvcc, etc. compilers as well.

**Comment by intractabilis** [comment]
Clang in fact supports it all. You can even compile it on Windows as fully compatible with cl. It's ironic you mentioned NVCC... it uses Clang front-end.

**Comment by sean-mcmanus** [comment]
There are many differences in the clang compiler that cause code to compile on clang but not gcc or msvc, etc. You can't configure clang to behave as specific versions of gcc or cl.exe.

**Comment by intractabilis** [comment]
We are not talking about compilers, we are talking about a C++ front-end you use as a library in your code for the purpose of IntelliSense. Clang parsing _API_ has enough flexibility to support anything. Most of differences between compilers anyway are managed by dumping all predefines with `-dM`.

**Comment by sean-mcmanus** [comment]
Yeah, I mean the front end compilers generate different errors or non-errors...at least in regards to C++20 in which the parsers don't implement everything 100% to standard yet. Not to mention it would require a lot of work to integrate the new parser. If we did decide to switch front-end compilers (which is unlikely) the change would have to come from the VS side, e.g. https://developercommunity.visualstudio.com/report?space=62&entry=suggestion, particularly since we're headed towards unifying the vcpkgsrv.exe and cpptools-srv binaries which provide IntelliSense.

**Comment by intractabilis** [comment]
Ok, got it, if we want to reproduce the exact error messages of a particular compiler, then it's a challenge. I agree that C++20 support is not quite uniform, though obviously EDG doesn't do a good job either, thus this bug is here. The only argument I have is that since 20M people are using cpptools, if the parser were open, it could have helped speeding up C++20 support in Clang.

Since the change has to come from the VS, I realize it hardly will ever happen.

(I cannot check the link without logging in).

**Comment by sean-mcmanus** [comment]
Well, the error doesn't need to be exact, but if certain code generates an error only with clang and not gcc then we wouldn't want to show an error for gcc users. Our C++20 support is still in progress. The root cause of https://developercommunity.visualstudio.com/t/C20-IntelliSense-vcpkgsrvexe-uses-e/1412745 may not be an EDG bug, but a bug in our code that our code that uses it, i.e. similar bugs could exist even if we switched to a different front-end parser. Bugs related to the range-v3 library have been fixed previously. 20 million people aren't using cpptools -- that number is just total installs.

**Comment by intractabilis** [comment]
> but if certain code generates an error only with clang and not gcc then we wouldn't want to show an error for gcc users

That is totally achievable. However, I agree that it may not worth the effort.

> may not be an EDG bug

Got it. That was my initial question.

> 20 million people aren't using cpptools -- that number is just total installs

Alright, let's divide it by 20, assuming a single person installed it 20 times. One million is still an incredible big amount of developers. Probably more than the number of swe in the whole FAANG?

**Comment by intractabilis** [comment]
Are the chances this will be resolved soon?

**Comment by sean-mcmanus** [comment]
Not sure yet...there is some unusually long delay in the Developer Community site so no internal bug has been created for the issue yet (even for a bug I created last Friday), but crashes and infinite loops are usually given higher priority. It won't be fixed for 1.4.0-insiders, but possibly 1.4.0-insiders2. There may also be some delay due to issues related to https://devblogs.microsoft.com/visualstudio/visual-studio-2022/

**Comment by intractabilis** [comment]
I think we can close this one.
